### PR TITLE
Update interval for arrow

### PR DESCRIPTION
**Pull request by handstuyennn**
This PR provides fix for issue duckdb/duckdb-wasm#1204 
@Mytherin

### CODE REVIEW

File: src/function/table/arrow_conversion.cpp [hunk 1]
```diff
@@ -509,6 +509,11 @@ void ColumnArrowToDuckDB(Vector &vector, ArrowArray &array, ArrowScanLocalState
 			IntervalConversionMonths(vector, array, scan_state, nested_offset, size);
 			break;
 		}
+		case ArrowDateTimeType::MONTH_DAY_NANO: {
+			FlatVector::SetData(vector, (data_ptr_t)array.buffers[1] + GetTypeIdSize(vector.GetType().InternalType()) *
```
Comment (Mytherin) [inline src/function/table/arrow_conversion.cpp:513]: Our interval type has microseconds, not nanoseconds, so directly using the interval type is unlikely to be correct. Could you add some tests for this using e.g. the Python client/integration?

File: tools/pythonpkg/src/python_objects.cpp [hunk 2]
```diff
@@ -403,9 +403,9 @@ py::object PythonObject::FromValue(const Value &val, const LogicalType &type) {
 	}
 	case LogicalTypeId::INTERVAL: {
 		auto interval_value = val.GetValueUnsafe<interval_t>();
-		uint64_t days = duckdb::Interval::DAYS_PER_MONTH * interval_value.months + interval_value.days;
-		return py::cast<py::object>(import_cache.datetime().timedelta()(
-		    py::arg("days") = days, py::arg("microseconds") = interval_value.micros));
+		return py::cast<py::object>(import_cache.pandas().DateOffset()(
+		    py::arg("months") = interval_value.months, py::arg("days") = interval_value.days,
+		    py::arg("nanoseconds") = interval_value.micros * Interval::MICROS_PER_MSEC));
```
Comment (pdet) [inline tools/pythonpkg/src/python_objects.cpp:408]: This creates a dependency on Pandas in our DuckDB -> Python code in a path unrelated to Pandas.
Why is timedelta insufficient here?

File: tools/pythonpkg/tests/fast/arrow/test_integration.py [hunk 3]
```diff
@@ -101,6 +103,41 @@ def test_decimals_roundtrip(self,duckdb_cursor):
         result = duckdb_conn.execute('select * from bigdecimal')
         assert (result.fetchone()[0] == 9999999999999999999999999999999999)
 
+    def test_intervals_roundtrip(self,duckdb_cursor):
```
Comment (pdet) [inline tools/pythonpkg/tests/fast/arrow/test_integration.py:106]: Can you also add a test where the interval is used in a nested structure?

File: tools/pythonpkg/tests/fast/test_result.py [hunk 4]
```diff
@@ -41,6 +41,10 @@ def test_result_timestamps(self, duckdb_cursor):
         assert rel.execute().fetchall() == [(datetime.datetime(2008, 1, 1, 0, 0, 11), datetime.datetime(2008, 1, 1, 0, 0, 1, 794000), datetime.datetime(2008, 1, 1, 0, 0, 1, 989260), datetime.datetime(2008, 1, 1, 0, 0, 1, 899268))]
 
     def test_result_interval(self):
+        try:
+            from pandas import DateOffset
+        except:
+            return
```
Comment (pdet) [inline tools/pythonpkg/tests/fast/test_result.py:47]: I don't think we have in our CI a scenario where pandas can't be installed.
But when skipping a test please use `pytest.importorskip()`

File: tools/pythonpkg/tests/fast/arrow/test_integration.py (hunk 3 above)
Comment (pdet) [reply tools/pythonpkg/tests/fast/arrow/test_integration.py:106]: And tests for boundary values? (i.e., the min and max value supported by arrow in the interval type)

Comment (pdet) [reply tools/pythonpkg/tests/fast/arrow/test_integration.py:106]: And null values

**Review by pdet — CHANGES_REQUESTED** [review]
Hi @handstuyennn thank you very much for your PR.

I requested some extra testing.

I'm also not a big fan of adding Pandas in the `FromValue` code since it adds a dependency on a code path that is independent of Pandas (Is used by our fetch code).

File: tools/pythonpkg/src/python_objects.cpp (hunk 2 above)
Comment (pdet) [reply tools/pythonpkg/src/python_objects.cpp:408]: @Mause maybe you have a suggestion for an interval representation that does not add a dependency?

Comment (david542542) [reply tools/pythonpkg/src/python_objects.cpp:408]: Here's a solution that doesn't require going outside python stdlib: https://stackoverflow.com/a/4131114/651174. Example solution copied below:

```
import datetime, calendar

def add_months(sourcedate, months):
    month = sourcedate.month - 1 + months
    year = sourcedate.year + month // 12
    month = month % 12 + 1
    day = min(sourcedate.day, calendar.monthrange(year,month)[1])
    return datetime.date(year, month, day)

somedate = datetime.date.today()
add_months(somedate,1)
# datetime.date(2023, 4, 3)
```

File: tools/pythonpkg/tests/fast/arrow/test_integration.py (hunk 3 above)
Comment (handstuyennn) [reply tools/pythonpkg/tests/fast/arrow/test_integration.py:106]: I updated test for above

**Comment by Mytherin** [comment]
Could you merge with master so the tests can run again?

File: tools/pythonpkg/src/python_objects.cpp (hunk 2 above)
Comment (Mause) [reply tools/pythonpkg/src/python_objects.cpp:408]: > @Mause maybe you have a suggestion for an interval representation that does not add a dependency?

I'm not aware of a standard library one aside from timedelta, but I see no issue with us defining our own? Either as named tuple or a dataclass

File: tools/pythonpkg/src/python_objects.cpp [hunk 5]
```diff
@@ -403,9 +403,9 @@ py::object PythonObject::FromValue(const Value &val, const LogicalType &type) {
 	}
 	case LogicalTypeId::INTERVAL: {
 		auto interval_value = val.GetValueUnsafe<interval_t>();
-		uint64_t days = duckdb::Interval::DAYS_PER_MONTH * interval_value.months + interval_value.days;
-		return py::cast<py::object>(import_cache.datetime().timedelta()(
-		    py::arg("days") = days, py::arg("microseconds") = interval_value.micros));
+		return py::cast<py::object>(import_cache.pandas().DateOffset()(
```
Comment (Mytherin) [inline tools/pythonpkg/src/python_objects.cpp:406]: This change seems unrelated to the Arrow change, no? Could we perhaps revert this change in this PR and potentially address this later?

Comment (handstuyennn) [reply tools/pythonpkg/src/python_objects.cpp]: Hi @Mytherin @pdet ,

I reverted code and updated tests. Please review it and give me any feedback

Thanks

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**